x-pack/plugins/cases/public/components/actions/tags/edit_tags_selectable.tsx

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useMemo, useReducer, useState, useEffect } from 'react';
+import type { EuiSelectableOption, IconType } from '@elastic/eui';
+import {
+  EuiSelectable,
+  EuiSpacer,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiText,
+  EuiButtonEmpty,
+  EuiHorizontalRule,
+  EuiIcon,
+  EuiHighlight,
+  EuiSelectableListItem,
+} from '@elastic/eui';
+
+import { FormattedMessage } from '@kbn/i18n-react';
+import { assertNever } from '@kbn/std';
+import { isEmpty } from 'lodash';
+import type { Case } from '../../../../common';
+import * as i18n from './translations';
+import type { TagsSelectionState } from './types';
+
+interface Props {
+  selectedCases: Case[];
+  tags: string[];
+  isLoading: boolean;
+  onChangeTags: (args: TagsSelectionState) => void;
+}
+
+type TagSelectableOption = EuiSelectableOption<{ tagIcon: IconType }>;
+
+const enum TagState {
+  CHECKED = 'checked',
+  PARTIAL = 'partial',
+  UNCHECKED = 'unchecked',
+}
+
+const enum Actions {
+  CHECK_TAG,
+  UNCHECK_TAG,
+}
+
+const enum ICONS {
+  CHECKED = 'check',
+  PARTIAL = 'asterisk',
+  UNCHECKED = 'empty',
+}
+
+type Action =
+  | { type: Actions.CHECK_TAG; payload: string[] }
+  | { type: Actions.UNCHECK_TAG; payload: string[] };
+
+interface Tag {
+  tagState: TagState;
+  dirty: boolean;
+  icon: IconType;
+}
+
+interface State {
+  tags: Record<string, Tag>;
+  tagCounterMap: Map<string, number>;
+}
+
+/**
+ * The EuiSelectable has two states values for its items: checked="on" for checked items
+ * and check=undefined for unchecked items. Given that our use case needs
+ * to track tags that are part in some cases and not part in some others we need
+ * to keep our own state and sync it with the EuiSelectable. Our state is always
+ * the source of true.
+ *
+ * In our state, a tag can be in one of the following states: checked, partial, and unchecked.
+ * A checked tag is a tag that is either common in all cases or has been
+ * checked by the user. A partial tag is a tag that is available is some of the
+ * selected cases and not available in others. A user can not make a tag partial.
+ * A unchecked tag is a tag that is either unselected by the user or is not available
+ * in all selected cases.
+ *
+ * State transitions:
+ *
+ * partial --> checked
+ * checked --> unchecked
+ * unchecked --> checked
+ *
+ * A dirty tag is a tag that the user clicked. Because the EuiSelectable
```

Comment:
Instead of using `dirty` could we rename it to `stateCommonForAllCases` or something like that?

To rename it I think we'd need to change this state slighly:

```
The ones that are not common in any of the selected tags are marked as unchecked and not dirty in our state and unchecked in EuiSelectable state.
```

Could we instead mark it as true initially? That would keep the state consistent with the name, aka

On initial load for tags that are not common in any of the selected cases (not as an actually object but just how each variable stated in this comment would be set).
```
{
  unchecked: true,
  stateCommonForAllCases: true,
  uncheckedInEuiSelectable: true,
}
```
 
Based on my reading of the code I don't _think_ that would affect the behavior and I think it would simplify the purpose of the `dirty` variable.